build_modules/lib/src/meta_module.dart

```diff
@@ -0,0 +1,433 @@
+import 'dart:math';
+
+import 'package:analyzer/analyzer.dart';
+import 'package:build/build.dart';
+import 'package:path/path.dart' as p;
+import 'package:json_annotation/json_annotation.dart';
+
+import 'modules.dart';
+
+part 'meta_module.g.dart';
+
+/// Returns the top level directory in [uri].
+///
+/// Throws an [ArgumentError] if [uri] is just a filename with no directory.
+String _topLevelDir(String uri) {
+  var parts = p.url.split(p.url.normalize(uri));
+  String error;
+  if (parts.length == 1) {
+    error = 'The uri `$uri` does not contain a directory.';
+  } else if (parts.first == '..') {
+    error = 'The uri `$uri` reaches outside the root directory.';
+  }
+  if (error != null) {
+    throw new ArgumentError(
+        'Cannot compute top level dir for path `$uri`. $error');
+  }
+  return parts.first;
+}
+
+/// Convert [importUri] found in [source] to an [AssetId], handling both
+/// `package:` imports and relative imports.
+///
+/// Returns null for `dart:` uris since they cannot be referenced properly
+/// by an `AssetId`.
+///
+/// Throws an [ArgumentError] if an [AssetId] can otherwise not be created. This
+/// might happen if:
+///
+/// * [importUri] is absolute but has a scheme other than `dart:` or `package:`.
+/// * [importUri] is a relative path that reaches outside of the current top
+///   level directory of a package (relative import from `web` to `lib` for
+///   instance).
+AssetId _importUriToAssetId(AssetId source, String importUri) {
+  var parsedUri = Uri.parse(importUri);
+  if (parsedUri.isAbsolute) {
+    switch (parsedUri.scheme) {
+      case 'package':
+        var parts = parsedUri.pathSegments;
+        var packagePath = p.url.joinAll(['lib']..addAll(parts.skip(1)));
+        if (!p.isWithin('lib', packagePath)) {
+          throw new ArgumentError(
+              'Unable to create AssetId for import `$importUri` in `$source` '
+              'because it reaches outside the `lib` directory.');
+        }
+        return new AssetId(parts.first, packagePath);
+      case 'dart':
+        return null;
+      default:
+        throw new ArgumentError(
+            'Unable to resolve import. Only package: paths and relative '
+            'paths are supported, got `$importUri`.');
+    }
+  } else {
+    // Relative path.
+    var targetPath =
+        p.url.normalize(p.url.join(p.url.dirname(source.path), parsedUri.path));
+    var dir = _topLevelDir(source.path);
+    if (!p.isWithin(dir, targetPath)) {
+      throw new ArgumentError(
+          'Unable to create AssetId for relative import `$importUri` in '
+          '`$source`  because it reaches outside the `$dir` directory.');
+    }
+    return new AssetId(source.package, targetPath);
+  }
+}
+
+/// An [AssetId] and all of its internal/external deps based on it's
+/// Directives.
+///
+/// Used to compute strongly connected components in the import graph for all
+/// "internal" deps. Any "external" deps are ignored during that computation
```

Comment:
So this is an interesting aspect of this algorithm that I hadn't thought about.

We should figure out a way to support modules which have a cycle containing files in more than one package. I think this is doable - with one requirement - which is that a a coarse-grained package can not depend on a fine-grained package.

We should talk offline about the strategy here.